File: leetcode/python/Reorder List/Reorder_List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a ListNode
    # @return nothing
    def reorderList(self, head):
        if head is None or head.next is None:
            return head
        slow = head
        fast = head
        while fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next
        fast = slow.next
        slow.next = None
        
        temp = fast.next
        fast.next = None
        while temp:
            next_node = temp.next
            temp.next = fast
            fast = temp
            temp = next_node
        
        tail = head
        while fast:
            next_node = fast.next
            fast.next = tail.next
            tail.next = fast
            tail = tail.next.next
            fast = next_node
        return head
            
        
from ListNode import ListNode
solu = Solution()
node1 = ListNode(1)
node2 = ListNode(2)
node3 = ListNode(3)
node4 = ListNode(4)
node5 = ListNode(5)
#node1.next = node2
node2.next = node3
node3.next = node4
#node4.next = node5
head = solu.reorderList(node1)
while head:
    print head.val
    head = head.next

File: leetcode/cpp/ReorderList.cpp
//
// Created by zhangqi on 2018/1/3.
// Time:  O(n)
// Space: O(1)
#include "ListNode.h"

class Solution {
public:
    void reorderList(ListNode* head) {
        auto slow = head;
        auto fast = head;
        if (!head){
            return;
        }
        while(fast->next && fast->next->next){
            slow = slow->next;
            fast = fast->next->next;
        }

        auto tmp = slow->next;
        slow->next = nullptr;
        slow = tmp;
        merge(head, reverse(slow));
    }

private :

    ListNode *reverse(ListNode *head){
        ListNode dummy{0};
        while (head){
            auto tmp = head->next;
            head->next = dummy.next;
            dummy.next = head;
            head = tmp;
        }
        return dummy.next;
    }

    ListNode *merge(ListNode* l1, ListNode* l2){
        ListNode dummy{0};
        auto ptr = &dummy;
        while (l1 && l2){
            auto tmp = l1->next;
            ptr->next = l1;
            ptr->next->next = l2;
            ptr = l2;

            l1 = tmp;
            l2 = l2->next;
        }
        if(l1){
            ptr->next = l1;
        }
        return dummy.next;
    }
};